Debugger improvements

LukasKalbertodt opened the pull request:
- Log and ASM views are now scrollable
- ASM view will cache decoded instructions to show instructions before PC
- Views are updated only when necessary, which should reduce CPU usage
- Better structure

CC #18 
Closes #26 
Closes #27

LukasKalbertodt committed:
Restructure code in the TUI debugger

LukasKalbertodt committed:
Wrap ASM view into scroll container

LukasKalbertodt committed:
Add `Word::map`

LukasKalbertodt committed:
Make log view scrollable

LukasKalbertodt committed:
Don't update debugger views when it's not necessary

LukasKalbertodt committed:
Improve ASM view

- Decoded instructions are cached
- Instructions before PC are showed too
- An arrow points to the instruction PC is pointing to

LukasKalbertodt committed:
Fix ASM view boot ROM cache invalidation

LukasKalbertodt committed:
Scroll ASM view to the current instruction

LukasKalbertodt committed:
Add getter methods for all registers of PPU

LukasKalbertodt committed:
Add second right panel for TUI debugger

LukasKalbertodt committed:
Make stack view scrollable

LukasKalbertodt committed:
Implement `Display` for `ppu::Phase`

LukasKalbertodt committed:
Fix a ton of bugs in PPU and teach PPU about disabled/enabled LCD

LukasKalbertodt committed:
Initialize the screen with a friendly blue to make debugging PPU easier

LukasKalbertodt committed:
Add big PPU debugging info box

LukasKalbertodt committed:
Reduce TUI FPS in pause mode to save CPU

LukasKalbertodt committed:
Show breakpoints in ASM view

LukasKalbertodt committed:
Fix forwarding of mouse events to children of the tab view

LukasKalbertodt committed:
Add and remove breakpoints via ASM view

LukasKalbertodt commented:
I couldn't keep myself from implementing more stuff. Most importantly:

- Fix quite a few PPU bugs (I know, this has nothing to do with the debugger, sorry for non-atomic PRs :P)
- Add box showing (almost) all PPU register (in a second right panel)
- Reduce CPU usage when debugger is paused
- You can now view, set and unset breakpoints by clicking in the ASM view

But I'm gonna stop now. So this PR is ready.

jovobe commented on core/src/machine/ppu.rs:
```diff
@@ -249,14 +309,21 @@ impl Ppu {
-                (43..144, _) => {
-                    // TODO: H-Blank
-                    // debug!("[ppu] hblank. current_line: {}", self.current_line);
+                (43..144, col) if self.phase() == Phase::HBlank && col == 160 => {
```
Use `SCREEN_WIDTH` instead of magic number `160`.

jovobe commented on core/src/machine/ppu.rs:
```diff
@@ -290,6 +357,13 @@ impl Ppu {
             self.push_pixel(display);
             pixel_pushed += 1;
+
+            // We we are at the end of the line, stop everything and go to
```
The "we" cloned itself 😱

jovobe commented on core/src/machine/ppu.rs:
```diff
@@ -290,6 +357,13 @@ impl Ppu {
+
+            // We we are at the end of the line, stop everything and go to
+            // H-Blank.
+            if self.current_column == 160 {
```
Use `SCREEN_WIDTH` instead of magic number `160`.

LukasKalbertodt committed:
Set current line to 0 when LCD gets disabled

LukasKalbertodt committed:
Address comments on PR

LukasKalbertodt committed:
Add CLI option to modify logger level and disable `trace!` in release builds

LukasKalbertodt committed:
Limit the number of log messages shown in the debug view

LukasKalbertodt committed:
Don't update window buffer when it didn't change

This reduces the time required for `NativeWindow::update()` from around
1,5ms to 200µs when the debugger is paused

LukasKalbertodt committed:
Save CPU by updating the TUI views less often (every 5th frame)

Now pausing and doing nothing in the debugger needs almost no
CPU. More importantly: since the TUI is only updated every fifth
frame, the CPU usage is also lower when the emulator is running.

LukasKalbertodt committed:
Accelerate rendering of the log view

Now we don't use `TextView` internally anymore and do everything
ourselves. This has a significant speed advantage, but means that
long lines aren't wrapping anymore. The log message can still use
line breaks, though.

LukasKalbertodt committed:
Add filtering for the log view

LukasKalbertodt committed:
Add generated comments to the ASM view that show register names

These comments can potentially show more information if we want to.

LukasKalbertodt commented:
I, again, worked on this instead of doing other stuff 😩

![image](https://user-images.githubusercontent.com/7419664/44818286-d50d9680-abe8-11e8-9575-96596dd7a434.png)

I guess these generated comments could come handy.

Other changes include log level via CLI and performance improvements.

LukasKalbertodt committed:
Make mahboi-desktop compile on Windows by disabling TUI debugger

There might be ways to make the debugger compile on Windows, but
I don't think it's worth the trouble.

jovobe commented on core/src/machine/ppu.rs:
```diff
@@ -184,20 +193,30 @@ impl Ppu {
+                        info!("[ppu] LCD was enabled");
+                        self.set_phase(Phase::OamSearch);
+                        self.cycle_in_line = 0;
+                        // TODO: also reset other stuff?
```
Do we need to reset the current line here too? (e.g. `self.current_line = Byte::new(0);`)

LukasKalbertodt replied:
I don't think so. The current line is set to 0 a few lines below, when the LCD is disabled. I actually think that most things should be "reset" when the LCD is disabled instead of when it is enabled. 

I think the TODO is a bit outdated. But it might still be useful until we figured this out.

jovobe commented:
I ignored everything in `desktop/src/debug` for the review.
